[MM-11876] Add cursor to posts list such as next_post_id and previous_post_id

saturninoabril opened the pull request:
#### Summary
Add cursor to posts list such as next_post_id and previous_post_id with the shape like:
```json
{
    "next_post_id": "post_id_3",
    "order": ["post_id_1", "post_id_2"],
    "posts": {
        "post_id_1": {...}, 
        "post_id_2": {...}, 
    },
    "previous_post_id": "post_id_0",
}
```

This adds cursor to posts list returned by the following:
- getPostsForChannel `GET api/v4/channels/{channel_id:[A-Za-z0-9]+}/posts`
  - [UPDATE] On query `?before={post_id}`, said before post ID is automatically set as `next_post_id`.  Only `previous_post_id` will be queried from the database after getting the post list.
  - [UPDATE] On query `?after={post_id}`, said after post ID is automatically set as `previous_post_id`.  Only `next_post_id` will be queried from the database after getting the post list.
  - [UPDATE] On query `?since={time}`, both `next_post_id` and `previous_post_id` will be queried separately from the database after getting the post list.
- getPostsForChannelAroundLastUnread `GET api/v4/users/{user_id:[A-Za-z0-9]+}/channels/{channel_id:[A-Za-z0-9]+}/posts/unread`
  - [UPDATE] both `next_post_id` and `previous_post_id` will be queried separately from the database after getting the post list.
  - ~~websocket event `WEBSOCKET_EVENT_POSTED = "posted"` will include `previous_post_id` on its published message~~ [UPDATE] will submit this separately

Note that cursor is NOT applied to the following:
- [UPDATE] getPost `GET api/v4/posts/{post_id}`
- getPinnedPosts `GET api/v4/channels/{channel_id:[A-Za-z0-9]+}/pinned`
- getFlaggedPostsForUser `GET api/v4/users/{user_id:[A-Za-z0-9]+}/posts/flagged`
- getPostThread `GET api/v4/channels/{channel_id:[A-Za-z0-9]+}/thread`
- searchPosts `GET api/v4/teams/{team_id:[A-Za-z0-9]+}/posts/search`

(Note I created this for I can't reopen https://github.com/mattermost/mattermost-server/pull/9342.  I guess I did the forced update first before reopening that.)

#### Ticket Link
Jira ticket: [MM-11876](https://mattermost.atlassian.net/browse/MM-11876)

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [ ] Added API documentation (to follow)
- [x] All new/modified APIs include changes to the drivers

saturninoabril commented:
@hmhealey I made the changes from your comment here - https://github.com/mattermost/mattermost-server/pull/9342

hmhealey commented on api4/post.go:
```diff
@@ -153,6 +156,9 @@ func getPostsForChannel(c *Context, w http.ResponseWriter, r *http.Request) {
 
 		list, err = c.App.GetPostsAfterPost(c.Params.ChannelId, afterPost, c.Params.Page, c.Params.PerPage)
+		if len(list.Order) > 0 {
+			previousPostId = afterPost
```
This only works if the page number is 0

saturninoabril replied:
Not sure if I understand correctly but this also works for page other than 0 with valid `afterPost` parameter.  I have a test that covered that.

hmhealey commented on api4/post.go:
```diff
@@ -161,6 +167,9 @@ func getPostsForChannel(c *Context, w http.ResponseWriter, r *http.Request) {
 
 		list, err = c.App.GetPostsBeforePost(c.Params.ChannelId, beforePost, c.Params.Page, c.Params.PerPage)
+		if len(list.Order) > 0 {
+			nextPostId = beforePost
```
Same here

saturninoabril replied:
Same with above explanation.

hmhealey commented on api4/post.go:
```diff
@@ -161,6 +167,9 @@ func getPostsForChannel(c *Context, w http.ResponseWriter, r *http.Request) {
+			nextPostId = beforePost
+		}
 	} else {
 		etag = c.App.GetPostsEtag(c.Params.ChannelId)
```
I think we can shortcut the next post id and previous post id here in a couple of cases:
1. If the page is 0, then the next post ID is empty
2. If `len(order) < c.Params.PerPage`, you're at the end of the channel, so the previous post ID is empty

saturninoabril replied:
We can't apply this assumption since the base reference is also `afterPost` or `beforePost` which may be in the middle of posts order.

saturninoabril replied:
Hmmm, I think it's applicable to `previousPostId` but not to `nextPostId`.  Let me check again.

saturninoabril replied:
Only applied to `previousPostId` to prevent unnecessary call to `GetPreviousPostIdFromPostList`.

hmhealey commented on api4/post.go:
```diff
@@ -185,6 +194,17 @@ func getPostsForChannel(c *Context, w http.ResponseWriter, r *http.Request) {
 		mlog.Error("Failed to prepare posts for getPostsForChannel response", mlog.Any("err", err))
 	}
 
+	if nextPostId == "" {
```
Similarly, we might want to differentiate here between the case that we know the next/previous post IDs are empty (ie you're at the start/end of the channel) so that we aren't doing this when we don't need to

saturninoabril replied:
As commented above, I only applied that to `previousPostId` but not to `nextPostId`.

hmhealey commented on app/post.go:
```diff
@@ -577,6 +577,58 @@ func (a *App) GetPostsAroundPost(postId, channelId string, offset, limit int, be
+	return result.Data.(*model.Post), nil
+}
+
+func (a *App) GetNextPostFromPostList(postList *model.PostList) string {
```
Can you rename this to `GetNextPostFromPostList` to make it clear that this returns an id and not the whole post?

hmhealey commented on app/post.go:
```diff
@@ -577,6 +577,58 @@ func (a *App) GetPostsAroundPost(postId, channelId string, offset, limit int, be
 	}
 }
 
+func (a *App) GetPostAfter(channelId string, time int64) (*model.Post, *model.AppError) {
```
Why use timestamps here instead of IDs? 0/5 between the two, but I initially expected that this to take a post ID

saturninoabril replied:
We're using timestamp at store to get post after, and since such timestamp is available before calling this function, we don't need to call get post just in case post ID is passed in here.
Maybe change function name, like `GetPostAfterTime`?

enahum replied:
Yeah just rename the function name to be more explicit, but it makes sense to query per timestamp

saturninoabril replied:
Rename the function to `GetPostBeforeTime` and `GetPostAfterTime`.

saturninoabril commented:
@hmhealey Thanks for the comments.  I made some changes, please take a look again.

saturninoabril commented:
Thanks @hmhealey for your offline check and explanation to me.  I have updated the logic in determining the next/previous post IDs.  Also updated the incorrect test case.
Please take a look again.

hmhealey commented on api4/post.go:
```diff
@@ -185,6 +185,29 @@ func getPostsForChannel(c *Context, w http.ResponseWriter, r *http.Request) {
+
+	var nextPostId string
+	if c.Params.Page == 0 {
+		nextPostId = beforePost
```
There's at least a few cases where this is incorrect. Say you have a channel with posts with IDs `post0` (the oldest) to `post9` (the newest), it won't work in at least the following cases:
1. When page is 0, it'll only set the next and previous post IDs from the before/after parameters, when it needs to get the other one from the database. For example, if you request `?after=post3&page=0&per_page=3`, you'll get
    ```
    {
        next_post_id: "", // This should be post7, but it's empty because we're looking for the `before` parameter
        order: ["post6", "post5", "post4"],
        prev_post_id: "post3"
    }
    ```
    The same thing happens with the previous post ID and `?before=...`.

hmhealey commented on api4/post.go:
```diff
@@ -185,6 +185,29 @@ func getPostsForChannel(c *Context, w http.ResponseWriter, r *http.Request) {
+	if c.Params.Page == 0 {
+		nextPostId = beforePost
+	} else if postListLength < c.Params.PerPage {
+		nextPostId = ""
```
2. If you're not getting the first page and less than a page of posts is returned, both the previous and next post IDs will be empty. For example, if you request `?before=post5&page=0&per_page=3` (and fix the first case), you'd get
    ```
    {
        next_post_id: "post5",
        order: ["post4", "post3", "post2"],
        prev_post_id: "post1"
    }
    ```
    but then when you get the next page with `?before=post5&page=1&per_page=3`, you'll get
    ```
    {
        next_post_id: "", // This should be "post2", but it's empty
        order: ["post1"],
        prev_post_id: ""
    }
    ```

saturninoabril commented:
Thanks Harrison for the detailed explanation.  Hmm, I guess I'm thinking differently here.  My confusion is coming from why do we need to know:
- `prevPostId` on `GetPostsAfter` other than when `page == 0`?  For me, `prevPostId` and `nextPostId` are necessary if `page == 0` and then only `nextPostId` on next page requests.
- same case for `nextPostId` on `GetPostsBefore`.

Anyway, I also understand why it should always be provided.
It's now updated.  Also, I separated the test for GetPostsAfter and GetPostsBefore.

saturninoabril commented on api4/post.go:
```diff
@@ -120,16 +120,20 @@ func getPostsForChannel(c *Context, w http.ResponseWriter, r *http.Request) {
+	if len(beforePost) > 0 && len(beforePost) != 26 {
+		c.SetInvalidParam("before")
+		return
+	}
```
Added validation on these query parameters so that it returns an error when passing invalid parameters.  This also avoids returning invalid next/post IDs.

saturninoabril commented:
@hmhealey @enahum Please take a look again, thanks

hmhealey commented:
Looks good now. Thanks for putting in so many tests
